### ISSUE TITLE
Aizynthfinder potential underperformance on ref-lin-600 and ref-lng-84

### ISSUE THREAD

**Issue by tagirshin** [ischemist/project-procrustes#95]
The problem is that the current AizynthFinder settings use the default configuration. For these benchmarks, the minimum number of steps is seven, but the AizynthFinder default is six. It's possible that the setting is overwritten somewhere in the project, but that makes it difficult to select configurations for SynPlanner for a fair comparison. If not overwritten, then there is a methodological problem with the comparison, and AizynthFinder is not in a fair position compared to other toolkits. 

So, my question is this intended behavior or a real problem?

```
search:
  iteration_limit: 100
expansion:
  uspto:
    - data/0-assets/model-configs/aizynthfinder/uspto_model.onnx
    - data/0-assets/model-configs/aizynthfinder/uspto_templates.csv.gz
  ringbreaker:
    - data/0-assets/model-configs/aizynthfinder/uspto_ringbreaker_model.onnx
    - data/0-assets/model-configs/aizynthfinder/uspto_ringbreaker_templates.csv.gz
filter:
  uspto: data/0-assets/model-configs/aizynthfinder/uspto_filter_model.onnx
stock:
  buyables-stock: data/1-benchmarks/stocks/buyables-stock.hdf5
  n1-n5-stock: data/1-benchmarks/stocks/n1-n5-stock.hdf5
  n5-stock: data/1-benchmarks/stocks/n5-stock.hdf5
```

**Comment by anmorgunov** [comment]
I didn't realize AiZynthFinder was limiting to 6 steps by default. I notice there's `max_transforms: 6` in the advanced use example, but I assumed the default would be something like no limit on max transforms. 

https://molecularai.github.io/aizynthfinder/configuration.html

I'll rerun AiZynthFinder on ref-lin and ref-lng as soon as possible. I'm extracting the planning scripts into a separate repo, so I opened an issue there: https://github.com/ischemist/project-pandora/issues/2.

I forgot to move the config files to that repo, will do soon.

> but that makes it difficult to select configurations for SynPlanner for a fair comparison

how so? I don't have any internal/private planning scripts, so anything you see here https://github.com/ischemist/project-pandora/tree/master/runtime/synplanner is what I used. 

Same with AiZynthFinder: there are no internal overwrites in https://github.com/ischemist/project-pandora/blob/master/scripts/aizynthfinder/3-run-aizyn-mcts.py.

The only question is whether `max_transforms` should be changed only for ref-lin-600 and ref-lng-84 or for all datasets. I'm inclined to set it to 10 (the max length of acceptable routes) for all benchmarks (because otherwise it'd be a potentially unfair comparison on runtime/computational cost). 

thanks for bringing this up

**Comment by anmorgunov** [comment]
i started a PR that will resolve this and #65 https://github.com/ischemist/project-pandora/pull/4

let me know if you have any thoughts on default for max_transforms and/or if I should pay attention to other config items like time_limit

**Comment by tagirshin** [comment]
> but that makes it difficult to select configurations for SynPlanner for a fair comparison
I forgot to mention that we are currently testing a new version of the SynPlanner 1.5.0 and we increased number of steps up to 10. And we got solvabilty up to 70% on this benchmark as we have included 10 steps as parameter (now we work to align evaluator to bring us closer to reference pathways, because on this side we are still bad).

I agree with you that, for MCTS-based planners, 10 transforms for linear and long routes is exactly what is needed. Regarding the time limit, there is also a question of whether we have to block Torch/TensorFlow threads. In general, when you run the benchmark on a remote machine with fixed resources, each tool is in the same conditions as the others. But maybe we need either to block explicitly torch/tensorflow resources (through env variables, see below) or run in docker containers so we sure we are controlling all the resources (RAM, num of proc, etc.). 

Also, in our update, we optimised operations to become faster, and even with Torch limitations on the number of threads, we can easily perform 1,000 iterations in under 20–30 seconds. In our benchmark (which we want to propose adding to the platform once we have reference pathways) we limit the time to 120 seconds, but the number of iterations is up to 100K. We also limit the tree size to 100,000 nodes (Syntheseus generates huge trees quite quickly; with 1 million nodes, for example, I get OOM errors). So I do not know how properly to cap algorithms (especially when each has separate view of what is iteration, step etc.). 

For now, I will create a separate issue inside project-pandora and put there all info from my local experiments to make comparisons as fair as possible.  

The example of env variables for torch:

```
os.environ.setdefault("OMP_NUM_THREADS", "6")
os.environ.setdefault("MKL_NUM_THREADS", "6")

import torch
torch.set_num_threads(6)
```

**Comment by anmorgunov** [comment]
> I will create a separate issue inside project-pandora

feel free to just open a PR updating synplanner exec scripts/configs in pandora.

> But maybe we need either to block explicitly torch/tensorflow resources or run in docker containers so we sure we are controlling all the resources (RAM, num of proc, etc.)

I understand the argument, and yes the Platonic ideal of fairness would necessitate it, but realistically, how an end user is going to use the planner? With some sophisticated docker setup OR in however way the docs/example use scripts are written?

This is why we label iteration_limit 100 as regular AiZynthFinder, even though in papers (e.g. PaRoutes) they use 500 (which we label as High effort). 

> In our benchmark (which we want to propose adding to the platform once we have reference pathways)

please feel free, I'm happy to host any benchmark on SynthArena

**Comment by anmorgunov** [comment]
https://github.com/ischemist/project-pandora/pull/4 is merged, now runner has cli --max-transforms arg. Will rerun the benchmarks soon.
